### PR TITLE
Destinations CDK: better integration tests

### PR DESCRIPTION
**Pull request by edgao**
Implement the test framework, write a minimal set of base tests, and implement those tests for destination-e2e-test. I stuck with our existing abstract class + concrete per-destination implementation strategy, because:
1. tests are opt-out
    1. like I mentioned in last (?) week's meeting - micronaut makes running the base class kind of wonky, so I think it's still best practice to have the `override testFoo() { super.testFoo() }` declarations. But even without that, we still get the test case.
2. it gives us an ok way to run individual test cases from inside intellij

general review guide:
1. DestinationMessage.kt has some simple convenience constructors
2. Start with IntegrationTest.kt, which is where the meat of the PR lives
3. Take a look at DestinationDataDumper, DestinationCleaner, ExpectedRecordMapper, and NameMapper - these are interfaces that destinations can/should implement. DataDumper doesn't have a noop implementation because every destination _must_ implement it; everything else is optional.
4. Check out DestinationProcess, which is how we launch the connector via micronaut (and eventually Docker). It's basically just a wrapper around CliRunner.
5. CheckIntegrationTest + BasicFunctionalityIntegrationTest show how to write some basic tests.
6. Then skim through all the destination-e2e-test code - it's hopefully pretty self-explanatory after everything else in this PR

### CODE REVIEW

**Comment by edgao** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45113?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#45113** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/45113?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#44874** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/44874?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @edgao and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: build.gradle
Comment (edgao) [inline build.gradle:58]: I'll revert this before merging, this was just driving me nuts locally

File: airbyte-integrations/connectors/destination-e2e-test/src/test-integration/java/io/airbyte/integrations/destination/e2e_test/TestingSilentDestinationAcceptanceTest.java
Comment (edgao) [inline airbyte-integrations/connectors/destination-e2e-test/src/test-integration/java/io/airbyte/integrations/destination/e2e_test/TestingSilentDestinationAcceptanceTest.java:21]: I didn't check what these two classes are doing, but will revive these tests before merging this PR

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt:63]: @johnny-schmidt before I go _too_ far down this path - this (+the related plumbing code in CliRunner + AirbyteConnectorRunner) seems like a nicer way to inject custom beans to the destination? I.e. this way, test authors don't need to define a new InputStreamFactory + do property dancing

but lmk if you have a different idea for doing this

Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt:64]: If the idea is that we pass in config & catalog explicitly, would it make more sense to provide the input data explicitly also? Just wrap it in an serializing InputStream?

Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationProcess.kt:64]: maaaybe, though I'm leaning towards no. There are tests that want to interleave messages and test logic, a la
```
sendSomeRecords()
verifySomething()
sendMoreRecords()
endSync()
verifySomethingElse()
```

At which point I'd rather wrap it in the destination process via the `sendMessage` function, rather than exposing the pipes to callers? I'm imagining something along these lines, which doesn't feel great?
```
val pipe = PipedOutputStream()
val pipeInputStream = PipedInputStream(pipe)
val dest = DestinationProcess(pipeInputStream)
messages1.each { pipe.write(serialize(it)) }
verifySomething()
messages2.each { pipe.write(serialize(it)) }
// flush all messages
pipe.close()
// this will magically cause the destination to shutdown
pipeInputStream.close()
```

plus I think we get most of the benefit by providing a utility method wrapping the destination process?
```
fun runSync(factory, config, catalog, messages) {
  val dest = factory.start(config, catalog)
  messages.each { dest.sendMessage(it) }
  return dest.readMessages()
}
```

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/write/BasicFunctionalityIntegrationTest.kt
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/write/BasicFunctionalityIntegrationTest.kt:58]: this test case is mostly here to (a) be a quick smoke test on our state handling, and (b) show the typical interaction with `DestinationProcess`. Most tests we'll actually write don't need any of the state stuff, and would look more like:
```kotlin
runSync(config, stream, listOf(records...))
dumpAndDiffRecords(...)
// (potentially more syncs, if we're e.g. testing refreshes)
```

File: airbyte-integrations/connectors/destination-e2e-test/build.gradle
Comment (edgao) [inline airbyte-integrations/connectors/destination-e2e-test/build.gradle:2]: this is a hack. We get the `integrationTestJava` task from this plugin. https://github.com/airbytehq/airbyte-internal-issues/issues/9864 will fix.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:49]: Does this mean that we never test the null namespace case?

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationCleaner.kt
Comment (johnny-schmidt) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationCleaner.kt:7]: Dq, why is the cleaner its own thing and not an entry point into the process?

**Review by johnny-schmidt — APPROVED** [review]
This looks solid to me. If you want to merge it down into e2e I'll start working on the final productionalization.

I think overall we'll want to iterate on these interfaces (mine and yours) and work out how we want to handle configs, injection in general, mapping to and from schemas and records, all that. But we can learn as we go.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt
Comment (edgao) [inline airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:38]: @johnny-schmidt / @stephane-airbyte I do want to call this piece out specifically - I copied this pattern from existing tests, but it could (/should) just be some utility class that gets constructed by test classes, instead of an extra layer of inheritance.

I'm basically reserving my next couple weeks to work on testing stuff (and probably won't have time to make big changes b/c OC)... so I'd prefer to just leave that debate for later, so we can at least get things merged + unblock further development? It should be a pretty simple switch, just a bit tedious.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:54]: we can still do that. Something along these lines, specifics of `config.setDefaultNamespace` tbd:
```kotlin
val modifiedConfig = config.setDefaultNamespace(randomizedNamespace)
runSync(modifiedConfig, records = listOf(Record(namespace=null, ...)))
verify(namespace=randomizedNamespace, ...)
```

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationCleaner.kt
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/DestinationCleaner.kt:7]: like why not have a `cleanup` method on DestinationProcess? I think they're pretty separate concerns - the dest process is just the connector (i.e. it supports spec/check/write)

but the cleaner is a weird out-of-connector thing. @stephane-airbyte 's recent Snowflake PR actually has a good example of what I want this class to do - https://github.com/airbytehq/airbyte/pull/45370/files#diff-c86b086ef3cc7bf564aedd699868dcc38661d51b3cc1bf34616720971866c6f6

(i.e. `show schemas` -> `drop schema`/etc)

**Comment by edgao** [comment]
> I think overall we'll want to iterate on these interfaces (mine and yours) and work out how we want to handle configs, injection in general, mapping to and from schemas and records, all that. But we can learn as we go.

+1. This is just the bare minimum to show what I want to do + get some basic demonstrations working, I fully expect we'll be tweaking/adding stuff as we go.

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt
Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:54]: though it's a bit trickier for s3, where namespace=null is an actual thing :/ I think we'd just disregard `randomizedNamespace` for that:
```kotlin
runSync(config, records=[Record(namespace = null, name = "test_stream_with_random_chars1234")])
```

(i.e. this field is purely advisory, it's up to each individual test to use it or not)

File: airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt
Comment (johnny-schmidt) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:38]: I'm fine with this as a foundation. We can talk through some of the implications of this stuff at the next CDK meeting?

Comment (edgao) [reply airbyte-cdk/bulk/core/load/src/testFixtures/kotlin/io/airbyte/cdk/test/util/IntegrationTest.kt:38]: sgtm!

**Pull request merged**
